bn_mp_radix_size_overestimate.c

```diff
@@ -0,0 +1,245 @@
+#ifdef MP_64BIT
+static const mp_digit logbases[65] = {
+#else
+static const mp_word logbases[65] = {
+#endif
+                      0ull,                    0ull,  2305843009213693952ull,
+    3654674702153732339ull,  4611686018427387904ull,  5354001665492895103ull,
+    5960517711367426291ull,  6473319721408181094ull,  6917529027641081856ull,
+    7309349404307464679ull,  7659844674706589055ull,  7976906213687625599ull,
+    8266360720581120243ull,  8532633055092329421ull,  8779162730621875046ull,
+    9008676367646627443ull,  9223372036854775808ull,  9425047617917838031ull,
+    9615192413521158631ull,  9795053960520401544ull,  9965687683920283007ull,
+   10127994423561913434ull, 10282749222901319551ull, 10430623713119086168ull,
+   10572203729794814195ull, 10708003330985790206ull, 10838476064306023373ull,
+   10964024106461197019ull, 11085005739835568998ull, 11201741516507896395ull,
+   11314519376860321395ull, 11423598928577852074ull, 11529215046068469760ull,
+   11631580915841357939ull, 11730890627131531983ull, 11827321386901076197ull,
+   11921035422734852583ull, 12012181624960263419ull, 12100896969734095496ull,
+   12187307757246061761ull, 12271530693133976959ull, 12353673836347420459ull,
+   12433837432775607386ull, 12512114650772492370ull, 12588592232115013503ull,
+   12663351069800359783ull, 12736466722332780120ull, 12808009872697200099ull,
+   12878046739008508147ull, 12946639442816362188ull, 13013846340199484158ull,
+   13079722320071570371ull, 13144319073519717325ull, 13207685337486676934ull,
+   13269867115674890971ull, 13330907879180520702ull, 13390848749049262950ull,
+   13449728662674133884ull, 13507584525721590347ull, 13564451351070528819ull,
+   13620362386074015347ull, 13675349229302017118ull, 13729441937791546026ull,
+   13782669125715645774ull, 13835058055282163712ull
+};
+#else
+#  ifdef _MSC_VER
+#    pragma message("mp_radix_size_overestimate: unknown or undefined MP_xBIT type")
+#  else
+#    warning "mp_radix_size_overestimate: unknown or undefined MP_xBIT type"
+#  endif
+#endif
+/* *INDENT-ON* */
+
+mp_err mp_radix_size_overestimate(const mp_int *a, const int radix, int *size)
+{
+   mp_int bi_bit_count, bi_k;
+#if ( (defined MP_8BIT) || (defined MP_16BIT) )
+   mp_int bi_k_bis;
+#endif
+   int bit_count;
+   mp_err err = MP_OKAY;
+
+   if (size == NULL) {
+      return MP_VAL;
+   } else {
+      *size = 0;
+   }
+
+   if ((radix < 2) || (radix > 64)) {
+      return MP_VAL;
+   }
+
+   if (a->used == 0) {
+      *size = 2;
+      return MP_OKAY;
+   }
+
+   bit_count = mp_count_bits(a);
+
+   /* A small shortcut for powers of two. */
+   if (!(radix&(radix-1))) {
+      unsigned int x = (unsigned int)radix;
+      int y, rem;
+      for (y=0; (y < 7) && !(x & 1u); y++) {
+         x >>= 1;
+      }
+      *size = bit_count/y;
+      rem = bit_count - ((*size) * y);
+      /* Add 1 for the remainder if any and 1 for "\0". */
+      *size += ((rem == 0) ? 1 : 2) + (a->sign == MP_NEG);
+      return MP_OKAY;
```

Comment:
Please remove rem here and just write + 3.

Reply:
> Please remove rem here and just write + 3.

That'll lengthen the error-bar to `-0, +3`.
No,  `-0, +1` is good and also expected.

But as I have more time now after the summer rush I will try my teeth at one of your suggestions and reduce the accuracy of the tables (or do something completely different, I have some ideas) to make it smaller and maybe even simpler for the cost of the size of `x` in `-0, +x`. There it will be where I can add your `3`.

Reply:
I don't see why you want this kind of precision? I don't expect anything from such a function except that it is overestimating and maybe that the error is a constant offset or a constant factors of the real size. This function is for allocation and if the function can be made simpler without being made worse we should do it. If you want to write another version which is more compact, I would appreciate that. But then the question is if we can then replace this function by the new function. In particular if you are specifing exact error bounds.

Reply:
> I don't see why you want this kind of precision?

I have it and was able to reach it without much ado.
Relaxing accuracy to save a dozen or so CPU cycles?
Reducing code complexity by a very small fraction (There is no "black art" in it)?

> But then the question is if we can then replace this function by the new function. In particular if you are specifing exact error bounds.

I don't know the error-bounds of the new function yet, too many variables. It will be most likely `-0, +6`  (including sign and remainder), although I aim for less

Reply:
Yes it is cheap but it is useless. The precision is not needed at all. +-1 is as good as +-10 or some other small constant.

Reply:
> +-1 is as good as +-10 or some other small constant.

:+1:

Reply:
If both of you are OK with such large error I can try something completely different here. Please give me a day or two for it.

Reply:
Cool! What about splitting your PR? One for the improvements to mp_radix_size by using ilogb and one for overestimate? The first one can be merged as is!